### PR TITLE
Add shared state, move 105 to own visitor file, move utility functionality out to new file visitor_utility.py

### PR DESCRIPTION
**Pull request by jakkdl**
Some housekeeping in preparation for #121 and future functionality that wants to use type tracking.

### CODE REVIEW

File: flake8_trio/visitors/flake8triovisitor.py [hunk 1]
```diff
@@ -124,22 +143,16 @@ def walk(self, *body: ast.AST) -> Iterable[ast.AST]:
...
-    def visit_Import(self, node: ast.Import):
-        for alias in node.names:
-            name = alias.name
-            if (
-                name in ("trio", "anyio")
-                and name not in self._library
-                and alias.asname is None
-            ):
-                self._library = self._library + (name,)
-
```
Comment (jakkdl) [inline flake8_trio/visitors/flake8triovisitor.py:136]: moved to `visitor_utility.py`

File: flake8_trio/visitors/helpers.py [hunk 2]
```diff
@@ -133,6 +139,17 @@ def has_exception(node: ast.expr) -> str | None:
...
+def is_nursery_call(node: ast.AST, name: str) -> bool:
+    assert name in ("start", "start_soon")
+    if isinstance(node, ast.Attribute):
+        if isinstance(node.value, ast.Name):
+            return node.attr == name and node.value.id.endswith("nursery")
+        if isinstance(node.value, ast.Attribute):
+            return node.attr == name and node.value.attr.endswith("nursery")
+    return False
+
+
```
Comment (jakkdl) [inline flake8_trio/visitors/helpers.py:152]: Moved from `visitors.py` as 105 got moved into it's own file.

File: flake8_trio/visitors/visitor105.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
+"""TRIO105: library async function must be immediately awaited."""
```
Comment (jakkdl) [inline flake8_trio/visitors/visitor105.py:1]: Entirely unchanged, moved out from `visitors.py` in preparation for getting a bunch of extra functionality.

File: flake8_trio/visitors/visitor2xx.py [hunk 4]
```diff
@@ -49,85 +49,6 @@ def visit_blocking_call(self, node: ast.Call):
 
 
 # used by Visitor212 and Visitor232
-class TypeTrackerVisitor200(Visitor200):
```
Comment (jakkdl) [inline flake8_trio/visitors/visitor2xx.py:52]: Moved to `visitor_utility` with some very minor changes.

File: flake8_trio/visitors/visitor2xx.py [hunk 5]
```diff
@@ -190,7 +111,7 @@ def visit_blocking_call(self, node: ast.Call):
 
 
 @error_class
-class Visitor212(TypeTrackerVisitor200):
+class Visitor212(Visitor200):
```
Comment (jakkdl) [inline flake8_trio/visitors/visitor2xx.py:114]: Now inherits from 200 like the rest of the 2xx so `visit_blocking_call` is called, and variable tracking is now shared across all runners.

File: flake8_trio/visitors/visitor_utility.py [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+from __future__ import annotations
+
+import ast
+
+from .flake8triovisitor import Flake8TrioVisitor
+from .helpers import utility_visitor
+
+
+@utility_visitor
+class VisitorTypeTracker(Flake8TrioVisitor):
```
Comment (jakkdl) [inline flake8_trio/visitors/visitor_utility.py:12]: Moved from `visitor2xx.py` with very minor changes.

File: flake8_trio/visitors/visitor_utility.py [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+        ):
+            self.variables[item.optional_vars.id] = vartype
+
+
+@utility_visitor
+class VisitorAwaitModifier(Flake8TrioVisitor):
+    def visit_Await(self, node: ast.Await):
+        if isinstance(node.value, ast.Call):
+            # add attribute to indicate it's awaited
+            setattr(node.value, "awaited", True)  # noqa: B010
```
Comment (jakkdl) [inline flake8_trio/visitors/visitor_utility.py:91]: Moved out from `runner.py`

File: flake8_trio/visitors/visitor_utility.py [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+        # specified with --anyio, otherwise assume Trio - but we update if we
+        # see imports
+        if self.options.anyio:
+            self.add_library("anyio")
+
+    def visit_Import(self, node: ast.Import):
+        for alias in node.names:
+            name = alias.name
+            if name in ("trio", "anyio") and alias.asname is None:
+                self.add_library(name)
```
Comment (jakkdl) [inline flake8_trio/visitors/visitor_utility.py:108]: moved out from `flake8triovisitor.py`

File: flake8_trio/visitors/visitors.py [hunk 9]
```diff
@@ -86,57 +87,6 @@ def visit_Yield(self, node: ast.Yield):
             self.error(node)
 
 
-# used in 105
```
Comment (jakkdl) [inline flake8_trio/visitors/visitors.py:89]: moved to `visitor105.py` / `helpers.py`

**Review by jakkdl — COMMENTED** [review]
Comments to ease review

File: flake8_trio/runner.py [hunk 10]
```diff
@@ -20,47 +21,48 @@
...
-        yield from runner._problems
+        yield from runner.state.problems
 
     def visit(self, node: ast.AST):
         """Visit a node."""
-        # don't bother visiting if no visitors are enabled, or all enabled visitors
-        # in parent nodes have marked novisit
-        if not self.visitors:
-            return
-
```
Comment (jakkdl) [inline flake8_trio/runner.py:53]: Variables can be *everywhere*!!

File: flake8_trio/runner.py [hunk 11]
```diff
@@ -20,47 +21,48 @@
...
         # tracks the subclasses that, from this node on, iterated through it's subfields
         # we need to remember it so we can restore it at the end of the function.
         novisit: set[Flake8TrioVisitor] = set()
 
         method = "visit_" + node.__class__.__name__
 
-        if m := getattr(self, method, None):
-            m(node)
```
Comment (jakkdl) [inline flake8_trio/runner.py:61]: The await transformer moved to `visitor_utility.py`, so runner no longer has any visitors itself to run. Also partly prep for libcst.

File: flake8_trio/runner.py [hunk 12]
```diff
@@ -89,10 +91,5 @@ def visit(self, node: ast.AST):
         self.visitors.update(novisit)
 
         # restore any outer state that was saved in the visitor method
-        for subclass in self.visitors:
+        for subclass in *self.utility_visitors, *self.visitors:
```
Comment (Zac-HD) [inline flake8_trio/runner.py:94]: I think this is an implicit tuple?  Would feel safer with parens around it though,

**Review by Zac-HD — APPROVED** [review]

File: flake8_trio/runner.py (hunk 12 above)
Comment (jakkdl) [reply flake8_trio/runner.py:94]: yep. I usually default to not putting explicit tuple parentheses unless necessary, but would maybe be nice to have them here.

**Pull request merged**